File: app/domains/acts/violation_fields.py
"""Декларативный контракт полей нарушения (#31A, бэкбон рефакторинга «Нарушения»).

Single source of truth для набора полей нарушения: ключ, метка, порядок
отображения, вид (`kind`) и два флага рендера (`small` — мелкий шрифт,
`show_label_in_preview` — показывать ли подпись поля в превью/форме). На этом
контракте позже стоит унификация подписей и единый рендер формы (следующие
задачи бэкбона); в ЭТОЙ задаче он только объявлен и закреплён стражами —
рендереры не трогаются.

Порядок `VIOLATION_FIELDS` — порядок полей в
``app.domains.acts.schemas.act_content.ViolationSchema`` (без `id`/`nodeId` —
это метаданные нарушения, не поля контента).

ВАЖНО: набор синхронизируется ВРУЧНУЮ с фронтовым зеркалом
``static/js/constructor/violation/violation-fields.js`` (как
``app/domains/acts/block_types.py`` ↔ ``static/js/constructor/block-types.js``,
``app/core/chat/names.py`` ↔ ``chat-client-actions.js``): фронт не импортирует
Python. Соответствие пиннится двумя тест-стражами —
``tests/domains/acts/test_violation_fields_guard.py`` (бэк) и
``tests/js/violation-fields.test.mjs`` (фронт, точные строки меток).

`small`: контракт ОПИСЫВАЕТ текущий рендер (позже DOCX/превью будут читать
размер отсюда). Сверено с ``formatters/docx/builders/violation.py``
(`build_violation`) и ``formatters/docx/styles.py``: 9pt-группа
(`Sizes.violation_pt`) — `violated` / `established` / `descriptionList` /
`additionalContent` (все ветки case/image-caption/freeText) → `small=True`.
`reasons` / `consequences` / `responsible` / `recommendations` рендерятся без
`size_pt` (дефолт `Sizes.body_pt`, 12pt; закреплено
`test_reasons_block_stays_12pt_non_italic`) → `small=False`.
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ViolationFieldDescriptor:
    """Описание одного поля нарушения: метка, порядок, вид, флаги рендера."""

    key: str
    label: str
    order: int
    kind: str  # "pair" | "list" | "additional" | "optional_text"
    small: bool
    show_label_in_preview: bool


VIOLATION_FIELDS: tuple[ViolationFieldDescriptor, ...] = (
    ViolationFieldDescriptor(
        key="violated", label="Нарушено", order=0, kind="pair",
        small=True, show_label_in_preview=True,
    ),
    ViolationFieldDescriptor(
        key="established", label="Установлено", order=1, kind="pair",
        small=True, show_label_in_preview=True,
    ),
    ViolationFieldDescriptor(
        # Заголовок убран (решение #12) — список описаний идёт без подписи.
        key="descriptionList", label="", order=2, kind="list",
        small=True, show_label_in_preview=False,
    ),
    ViolationFieldDescriptor(
        key="additionalContent", label="", order=3, kind="additional",
        small=True, show_label_in_preview=False,
    ),
    ViolationFieldDescriptor(
        key="reasons", label="Причины", order=4, kind="optional_text",
        small=False, show_label_in_preview=True,
    ),
    ViolationFieldDescriptor(
        key="consequences", label="Последствия", order=5, kind="optional_text",
        small=False, show_label_in_preview=True,
    ),
    ViolationFieldDescriptor(
        # Канон #11: "Ответственные" (не "Ответственный", как в DOCX-builder'е —
        # выравнивание подписи форматтеров будет отдельной задачей бэкбона).
        key="responsible", label="Ответственные", order=6, kind="optional_text",
        small=False, show_label_in_preview=True,
    ),
    ViolationFieldDescriptor(
        key="recommendations", label="Рекомендации", order=7, kind="optional_text",
        small=False, show_label_in_preview=True,
    ),
)

LABELS: dict[str, str] = {field.key: field.label for field in VIOLATION_FIELDS}

# Подпись кейса дополнительного контента ("Кейс 1", "Кейс 2", ...).
CASE_LABEL_TEMPLATE = "Кейс {n}"

# Свободный текст дополнительного контента — без подписи (решение #10).
FREE_TEXT_LABEL = ""

File: static/js/constructor/violation/violation-fields.js
/**
 * Декларативный контракт полей нарушения (#31A, бэкбон рефакторинга «Нарушения»).
 *
 * Зеркало реестра описаний полей нарушения: ключ, метка, порядок, вид (kind)
 * и два флага рендера (small — мелкий шрифт, showLabelInPreview — показывать
 * подпись поля в превью/форме). На этом контракте позже стоит унификация
 * подписей и единый рендер формы (следующие задачи бэкбона) — в этом файле
 * контракт только объявлен, рендереры не трогаются.
 *
 * ВАЖНО: набор синхронизируется ВРУЧНУЮ с бэкенд-реестром
 * app/domains/acts/violation_fields.py (как block_types.py ↔ block-types.js):
 * бэк не импортирует JS. Точные строки меток и порядок закреплены
 * тест-стражем tests/js/violation-fields.test.mjs; соответствие схеме
 * ViolationSchema — tests/domains/acts/test_violation_fields_guard.py.
 */

export const VIOLATION_FIELDS = Object.freeze([
  Object.freeze({ key: 'violated', label: 'Нарушено', order: 0, kind: 'pair', small: true, showLabelInPreview: true }),
  Object.freeze({ key: 'established', label: 'Установлено', order: 1, kind: 'pair', small: true, showLabelInPreview: true }),
  Object.freeze({ key: 'descriptionList', label: '', order: 2, kind: 'list', small: true, showLabelInPreview: false }),
  Object.freeze({ key: 'additionalContent', label: '', order: 3, kind: 'additional', small: true, showLabelInPreview: false }),
  Object.freeze({ key: 'reasons', label: 'Причины', order: 4, kind: 'optional_text', small: false, showLabelInPreview: true }),
  Object.freeze({ key: 'measures', label: 'Принятые меры', order: 5, kind: 'optional_text', small: false, showLabelInPreview: true }),
  Object.freeze({ key: 'consequences', label: 'Последствия', order: 6, kind: 'optional_text', small: false, showLabelInPreview: true }),
  Object.freeze({ key: 'responsible', label: 'Ответственные', order: 7, kind: 'optional_text', small: false, showLabelInPreview: true }),
]);

export const VIOLATION_LABELS = Object.freeze(
  Object.fromEntries(VIOLATION_FIELDS.map(f => [f.key, f.label]))
);

// Подпись кейса дополнительного контента ("Кейс 1", "Кейс 2", ...).
export const CASE_LABEL_TEMPLATE = 'Кейс {n}';

// Свободный текст дополнительного контента — без подписи (решение #10).
export const FREE_TEXT_LABEL = '';

// Window-globals для совместимости с inline-скриптами в шаблонах.
if (typeof window !== 'undefined') {
  window.VIOLATION_FIELDS = VIOLATION_FIELDS;
  window.VIOLATION_LABELS = VIOLATION_LABELS;
  window.CASE_LABEL_TEMPLATE = CASE_LABEL_TEMPLATE;
  window.FREE_TEXT_LABEL = FREE_TEXT_LABEL;
}
